context/src/main/java/io/micronaut/runtime/beans/DefaultIntrospectionBeanMapHandler.java

```diff
@@ -0,0 +1,305 @@
+        return map(input, mapStrategy, left, outputIntrospection);
+    }
+
+    @Override
+    public <I, O> O map(I input, BeanMapper.MapStrategy mapStrategy, BeanIntrospection<I> inputIntrospection, BeanIntrospection<O> outputIntrospection) {
+        if (mapStrategy == null) {
+            mapStrategy = BeanMapper.MapStrategy.DEFAULT;
+        }
+        @NonNull Argument<?>[] constructorArguments = outputIntrospection.getConstructorArguments();
+        BeanMapper.MapStrategy.ConflictStrategy conflictStrategy = mapStrategy.conflictStrategy();
+        if (constructorArguments.length == 0) {
+            O output = outputIntrospection.instantiate();
+            if (input != null) {
+                return map(input, output, mapStrategy, inputIntrospection, outputIntrospection);
+            }
+            return output;
+        } else {
+            Objects.requireNonNull(input, "Input cannot be null");
+            // constructor binding
+            Object[] parameters = new Object[constructorArguments.length];
+            for (int i = 0; i < constructorArguments.length; i++) {
+                Argument<?> constructorArgument = constructorArguments[i];
+                BeanProperty<I, ?> property = inputIntrospection.getProperty(constructorArgument.getName(), constructorArgument.getType()).orElse(null);
+                if (property != null) {
+                    Object v = property.get(input);
+                    if (v != null) {
+                        parameters[i] = v;
+                    } else if (constructorArgument.isDeclaredNonNull()) {
+                        throw new IllegalArgumentException(PREFIX_NON_NULL_ARG + constructorArgument.getName());
+                    }
+                } else if (constructorArgument.isDeclaredNonNull()) {
+                    throw new IllegalArgumentException(PREFIX_NON_NULL_ARG + constructorArgument.getName());
+                }
+            }
+            O output = outputIntrospection.instantiate(parameters);
+            return map(input, output, mapStrategy, inputIntrospection, outputIntrospection);
+        }
+    }
+
+    @Override
+    public <O> @NonNull O map(Map<String, Object> input, Class<O> outputType, BeanMapper.MapStrategy mapStrategy) {
+        Objects.requireNonNull(outputType, "Output type cannot be null");
+        BeanIntrospection<O> right = beanIntrospector.getIntrospection(outputType);
+        return map(input, mapStrategy, right);
+    }
+
+    @Override
+    public <O> O map(Map<String, Object> input, BeanMapper.MapStrategy mapStrategy, BeanIntrospection<O> outputIntrospection) {
+        if (mapStrategy == null) {
+            mapStrategy = BeanMapper.MapStrategy.DEFAULT;
+        }
+        BeanMapper.MapStrategy.ConflictStrategy conflictStrategy = mapStrategy.conflictStrategy();
+        @NonNull Argument<?>[] constructorArguments = outputIntrospection.getConstructorArguments();
+        if (constructorArguments.length == 0) {
```

Comment:
yes I agree, I actually have in mind a new API because I want to add generic support for the `Builder` pattern as well and if we had `BeanIntrospection.builder() -> BeanBuilder` we could abstract constructor arguments.